concrete/authentication/external_concrete5/controller.php

```diff
@@ -0,0 +1,187 @@
+     */
+    public function saveAuthenticationType($args)
+    {
+        $passedUrl = trim($args['url']);
+
+        if ($passedUrl) {
+            try {
+                $url = Url::createFromUrl($passedUrl);
+
+                if (!(string)$url->getScheme() || !(string)$url->getHost()) {
+                    throw new \InvalidArgumentException('No scheme or host provided.');
+                }
+
+            } catch (\Exception $e) {
+                throw new \InvalidArgumentException('Invalid URL.');
+            }
+        }
+
+        $config = $this->app->make(Repository::class);
+        $config->save('auth.external_concrete5.url', $args['url']);
+        $config->save('auth.external_concrete5.appid', $args['apikey']);
+        $config->save('auth.external_concrete5.secret', $args['apisecret']);
+        $config->save('auth.external_concrete5.registration.enabled', (bool) $args['registration_enabled']);
+        $config->save('auth.external_concrete5.registration.group', intval($args['registration_group'], 10));
+    }
+
+    /**
+     * Controller method for type_form
+     * This method is called just before rendering type_form.php, use it to set data for that template
+     */
+    public function edit()
+    {
+        $config = $this->app->make(Repository::class);
+        $this->set('form', $this->app->make('helper/form'));
+        $this->set('data', $config->get('auth.external_concrete5', ''));
+        $this->set('redirectUri', $this->urlResolver->resolve(['/ccm/system/authentication/oauth2/external_concrete5/callback']));
+
+        $list = $this->app->make(GroupList::class);
+        $list->includeAllGroups();
+        $this->set('groups', $list->getResults());
+    }
+
+    /**
+     * Controller method for form
+     * This method is called just before form.php is rendered, use it to set data for that template
+     */
+    public function form()
+    {
+        $this->setData();
+    }
+
+    /**
+     * Controller method for the hook template
+     * This method is called before hook.php is rendered, use it to set data for that template
+     */
+    public function hook()
+    {
+        $this->setData();
+    }
+
+    /**
+     * Controller method for the hooked template
+     * This method gets called before hooked.php is rendered, use it to set data for that template
+     */
+    public function hooked()
+    {
+        $this->setData();
+    }
+
+    /**
+     * Method for setting general data for all views
+     */
+    private function setData()
+    {
+        $authUrl = $this->urlResolver->resolve(['/ccm/system/authentication/oauth2/external_concrete5/attempt_auth']);
+        $attachUrl = $this->urlResolver->resolve(['/ccm/system/authentication/oauth2/external_concrete5/attempt_attach']);
+        $baseUrl = $this->urlResolver->resolve(['/']);
+        $path = $baseUrl->getPath();
+        $path->remove('index.php');
+        $name = trim((string) array_get($data, 'name', t('External concrete5')));
```

Comment:
I'm confused about where `$data` is set, am I missing something?

Reply:
Whoops, missed the definition :)